Zomes page (#511)
holochain/docs-pages: src/pages/build/application-structure.md

* It's small and fast --- it can get compiled to machine code for near-native speed.
* Holochain is written in Rust, and Rust has an excellent WebAssembly engine called [Wasmer](https://wasmer.io/) that works on all the major operating systems.
* It provides a secure sandbox to run untrusted code within.

!!!

A zome has access to Holochain's host API and also exposes functions of its own. Some of these functions are **required callbacks** and some of them you invent yourself to create your back end's API.

There are two kinds of zome:

* An **integrity zome** defines a set of data types --- your application's schema --- and validation rules for operations that create, update, or delete data of those types. If you're used to thinking of apps in terms of [model-view-controller (MVC)](https://en.wikipedia.org/wiki/Model%E2%80%93view%E2%80%93controller), it's like your application's model.
* A **coordinator zome** defines a set of functions for interacting with data and peers. You can think of it as your application's controller.

> Zomes are usually created as pairs --- an integrity zome that defines a portion of your schema and a coordinator zome that defines <abbr title="create, read, update, delete">CRUD</abbr> functions for that schema. You don't have to do it this way though; coordinator zomes don't need an integrity zome if they don't manipulate data, or they can depend on multiple integrity zomes, or multiple coordinators can depend on the same integrity zome.

Reviewer: ```suggestion
Zomes are usually created as pairs --- an integrity zome that defines a portion of your data types and rules, and a coordinator zome that defines functions that operate on those data types. You don't have to do it this way though; coordinator zomes don't need an integrity zome if they don't manipulate data, or they can depend on multiple integrity zomes, or multiple coordinators can depend on the same integrity zome.
```

A schema suggests static rules defined in structured data to me. We usually talk about "rules of the game" with integrity zomes and I think "rules" better suggest that you define your own rules in code.

For coordinators, you are not limited to defining simple CRUD against the data. You could write your coordinator that way, but I think it's a mistake to repeatedly cross the WASM boundary for CRUD and have all your logic on the front end. I think it makes far more sense to have your backend doing more creative work. Like create_entry + create_link, is already doing more than a "C" and the two operations will succeed or fail together. That's quite important.

Author: oh, that's kind of what I was glomming into the definition of 'CRUD' -- the creative stuff. Ditto re: schema -- I see it as an imperative schema where devs are used to writing declarative schemas. But what I'm getting here is that these words are conveying things that aren't accurate, so I'll accept the suggestion.

> Zomes are usually created as pairs --- an integrity zome that defines a portion of your schema and a coordinator zome that defines <abbr title="create, read, update, delete">CRUD</abbr> functions for that schema. You don't have to do it this way though; coordinator zomes don't need an integrity zome if they don't manipulate data, or they can depend on multiple integrity zomes, or multiple coordinators can depend on the same integrity zome.

Reviewer 2: ```suggestion
Zomes are usually created as pairs --- an integrity zome that defines your schema and validations rules, along with a coordinator zome that defines <abbr title="create, read, update, delete">CRUD</abbr> functions for that schema. You don't have to do it this way though; coordinator zomes don't need an integrity zome if they don't manipulate data, or they can depend on multiple integrity zomes, or multiple coordinators can depend on the same integrity zome.
```

Reviewer 2: I deleted the "portion" here because it's confusing in this context.  maybe it could come back but it would need  "your app's schema"

> If you mean for your zomes to be reused by other projects, you can share them via a public Git repository or [crates.io](https://crates.io) (tag your crates with `#holochain` so others can find them).

Reviewer: Is anybody using `crates.io`? I haven't come across that yet. I'm assuming there's some reason people are using `git` dependencies. Any insight here @matthme or @zippy?

Author: it'd be good to pave the cowpaths, so I'll defer to what's actually being done in practice.

Reviewer 2: Joel has used it for utilities crates:  https://crates.io/crates/hc_iz_membrane_manager

[Read more on the Zomes page](/build/zomes/).

### DNA

> One or more zomes are bundled into a **DNA**. When two or more participants install and run a DNA, an isolated peer-to-peer network is created among them to interact and store shared data.

Reviewer: ```suggestion
One or more zomes are bundled into a **DNA**, including at least one integrity zome. When two or more participants install and run a DNA, a new peer-to-peer network is created among them to interact and store shared data.
```

Reviewer: An "isolated" network has an existing meaning that we do not meet https://www.meter.com/resources/what-is-network-isolation

Author: On a casual reading it seems like it meets the meaning perfectly -- what nuance am I missing?

> One or more zomes are bundled into a **DNA**. When two or more participants install and run a DNA, an isolated peer-to-peer network is created among them to interact and store shared data.

Reviewer 2: Here you use the term "Particpant" elsewhere I see "Peer".  This should be consistent.

Author: I've been using different terms for differnt situations:

* 'Participant' as the general hand-wavey term for user/agent/whatever
* 'Peer' as a 'someone else' in relation to a primary participant in an interaction -- the Bob to an Alice
* 'Agent' as the technical term for someone/thing in control of a cell
* 'User' when I can't think of a better term for 'human operator of a cell'

How do those feel? And have I been using them inconsistently? Sometimes it's a struggle to hit on the best term in a given context. I could also just define 'agent' at some root-level documentation and use it consistently everywhere, although sometimes the more specific term really is the best one.

**A DNA, and the network created for it, is uniquely defined by its integrity zomes, plus any modifiers.** The hash of the integrity zomes plus modifiers is called the **DNA hash**, and is the unique identifier for the network.

!!! info Why not coordinator zomes?

> Coordinator zomes are bundled with a DNA, but they don't contribute to its uniqueness. This lets you hot-swap coordinators as you fix bugs and add features, without causing the DNA to **fork** a new network. The only things that should cause a fork are changes to integrity code --- the 'rules of the game' for the participants.

Reviewer: This is good but I think we want to talk about forking separately. That's an option, but the concept of migration is different to a fork. I don't know what the best thing to say here would be because you currently can't do DNA migration.

Maybe we should be direct and say that for now the lifecycle of data is completely tied to the DNA until we implement migrations?

!!!

> Because each DNA has its own isolated peer network and data store, you can use the DNA concept to come up with creative approaches to privacy and access, separation of responsibilities, or data retention.

Reviewer: There is no privacy without encryption, DHT data is not protected by anything Holochain does. Its *integrity* is protected, but read access to it is not

### hApp

> Multiple DNAs come together in a **hApp** (Holochain app). Each DNA fills a named **role** in the hApp, and can be seen as a [microservice](https://en.wikipedia.org/wiki/Microservices).

Reviewer: That's a nice analogy, I like that

Each peer generates its own public/private key pair when they install a hApp. Their public key acts as their **agent ID** which identifies them as a participant in all the networks created for all the DNAs in the hApp. When a DNA is activated, it's bound to this key pair and becomes a DNA instance or **cell**.

The hApp can specify a few provisioning strategies for its DNAs:

* A cell can be instantiated at app installation time.
* A new cell can be **cloned** from an existing DNA at any time _after the hApp is installed_, with an optional limit on the number of clones.

> A hApp can optionally include a web-based UI that supporting Holochain runtimes <!-- TODO: link --> can serve to the user.

Reviewer: I think `runtimes` deserves its own introduction/definition. I think this term is not accurate to the standard meaning of the term and is something we've fallen into using internally. I'd probably put it in bold like the other Holochain defined terms and give it a quick explanation

Reviewer 2: I think we should just define this by example in our case and point to various runtimes (whenever we actually add this).

!!! info A hApp always runs locally

The big difference with peer-to-peer stacks like Holochain is that **all the code** --- both the back end and the front end --- **runs on the devices of the participants themselves**.

That means that a DNA doesn't exist as some piece of code that runs 'out there somewhere' --- instead it runs from the perspective of an individual. DNA + agent = cell.

> There can still be bots or system-level services that do automated tasks. Those functions just have to be handled by one of the peers, and that peer doesn't have to be a human.

Reviewer: Might be nice to tie back to the "agent" term here? There's a reason we use agent to cover the actor and it's this
